File: packages/core/src/firefly.ts
import { clamp01, type DayNumber } from "./types.js";

/** Năng lượng Đom giảm ~0.34 mỗi ngày không học (kế thừa từ app SwiftUI). */
export const FIREFLY_DECAY_PER_DAY = 0.34;

export interface FireflyState {
  energy: number; // 0..1
  lastUpdatedDay: DayNumber;
}

/** Năng lượng hiện tại sau khi trừ phần đã suy giảm tới `today`. */
export function currentEnergy(state: FireflyState, today: DayNumber): number {
  const idleDays = Math.max(0, today - state.lastUpdatedDay);
  return clamp01(state.energy - FIREFLY_DECAY_PER_DAY * idleDays);
}

/** Cho Đom "ăn" — nạp năng lượng khi học (amount 0..1). */
export function feed(
  state: FireflyState,
  amount: number,
  today: DayNumber
): FireflyState {
  const base = currentEnergy(state, today);
  return { energy: clamp01(base + amount), lastUpdatedDay: today };
}

/** Mức sáng hiển thị (0..1) — dùng cho hiệu ứng nhấp nháy. */
export function glowIntensity(state: FireflyState, today: DayNumber): number {
  return currentEnergy(state, today);
}

File: apps/ios/DomDom/Core/Firefly.swift
import Foundation

/// Port của @domdom/core (firefly.ts). TypeScript là chuẩn — giữ công thức khớp.
enum Firefly {
    static let decayPerDay = 0.34

    static func currentEnergy(energy: Double, lastUpdatedDay: Int, today: Int) -> Double {
        let idleDays = max(0, today - lastUpdatedDay)
        return clamp01(energy - decayPerDay * Double(idleDays))
    }

    static func feed(energy: Double, amount: Double) -> Double {
        clamp01(energy + amount)
    }

    private static func clamp01(_ x: Double) -> Double {
        max(0, min(1, x))
    }
}
